File: ug-scraper/internal/api/handlers/onsong_cloud.go
package handlers

import (
	"bytes"
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
	"net/textproto"
	"os"
	"time"

	"github.com/gofiber/fiber/v2"
)

const onsongCloudURL = "https://onsongapp.com/drive/files/~/"

// OnSongCloudHandler handles uploads to OnSong Cloud Drive
type OnSongCloudHandler struct {
	token string
}

// NewOnSongCloudHandler creates a new OnSong Cloud handler using the ONSONG_TOKEN env var
func NewOnSongCloudHandler() *OnSongCloudHandler {
	return &OnSongCloudHandler{
		token: os.Getenv("ONSONG_TOKEN"),
	}
}

// GetConfig returns whether the OnSong Cloud token is configured
func (h *OnSongCloudHandler) GetConfig(c *fiber.Ctx) error {
	return c.JSON(fiber.Map{
		"configured": h.token != "",
	})
}

// Send uploads a tab as a .txt file to OnSong Cloud Drive
func (h *OnSongCloudHandler) Send(c *fiber.Ctx) error {
	if h.token == "" {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"error": "OnSong Cloud token not configured",
		})
	}

	var req struct {
		Title   string `json:"title"`
		Artist  string `json:"artist"`
		Content string `json:"content"`
	}

	if err := c.BodyParser(&req); err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"error":   "invalid request body",
			"details": err.Error(),
		})
	}

	if req.Title == "" || req.Content == "" {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"error": "title and content are required",
		})
	}

	filename := fmt.Sprintf("%s - %s.txt", req.Title, req.Artist)
	fmt.Printf("\n☁️ Uploading to OnSong Cloud: %s\n", filename)

	// Build multipart body
	var body bytes.Buffer
	writer := multipart.NewWriter(&body)

	// Create form file part with explicit content-type
	partHeader := make(textproto.MIMEHeader)
	partHeader.Set("Content-Disposition", fmt.Sprintf(`form-data; name="file[0]"; filename="%s"`, filename))
	partHeader.Set("Content-Type", "text/plain")

	part, err := writer.CreatePart(partHeader)
	if err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"error":   "failed to create multipart form",
			"details": err.Error(),
		})
	}

	if _, err := io.WriteString(part, req.Content); err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"error":   "failed to write content",
			"details": err.Error(),
		})
	}
	writer.Close()

	httpReq, err := http.NewRequest("POST", onsongCloudURL, &body)
	if err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"error":   "failed to create request",
			"details": err.Error(),
		})
	}

	httpReq.Header.Set("Authorization", h.token)
	httpReq.Header.Set("Content-Type", writer.FormDataContentType())
	httpReq.Header.Set("Accept", "application/json")

	client := &http.Client{Timeout: 30 * time.Second}
	resp, err := client.Do(httpReq)
	if err != nil {
		fmt.Printf("❌ OnSong Cloud upload failed: %v\n\n", err)
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"error":   "upload failed",
			"details": err.Error(),
		})
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		bodyBytes, _ := io.ReadAll(resp.Body)
		fmt.Printf("❌ OnSong Cloud upload failed: status=%d body=%s\n\n", resp.StatusCode, string(bodyBytes))
		return c.Status(fiber.StatusBadGateway).JSON(fiber.Map{
			"error":   "OnSong Cloud returned an error",
			"status":  resp.StatusCode,
			"details": string(bodyBytes),
		})
	}

	fmt.Printf("✅ OnSong Cloud upload successful: %s\n\n", filename)
	return c.JSON(fiber.Map{
		"success":  true,
		"filename": filename,
	})
}

File: ug-scraper-v1/packages/backend/src/routes/onsong-cloud.ts
/**
 * OnSong Cloud Drive integration routes.
 *
 * GET  /api/onsong/config  — returns { configured: boolean }
 * POST /api/onsong/send    — uploads a song as a .onsong file to OnSong Cloud Drive
 *
 * Auth: the raw ONSONG_TOKEN value is sent as the Authorization header (no Bearer prefix).
 * This matches the OnSong Cloud Drive API contract.
 */

import type { FastifyInstance } from 'fastify';
import { z } from 'zod';

const ONSONG_UPLOAD_URL = 'https://onsongapp.com/drive/files/~/';

const SendBodySchema = z.object({
  title: z.string().min(1),
  artist: z.string().min(1),
  content: z.string().min(1),
});

export async function onsongCloudRoutes(app: FastifyInstance): Promise<void> {
  /**
   * GET /api/onsong/config
   * Returns { configured: true } when ONSONG_TOKEN is set, { configured: false } otherwise.
   */
  app.get('/api/onsong/config', async (_request, reply) => {
    const token = process.env.ONSONG_TOKEN ?? '';
    return reply.send({ configured: token.length > 0 });
  });

  /**
   * POST /api/onsong/send
   * Uploads the provided OnSong content to OnSong Cloud Drive.
   * Body: { title: string, artist: string, content: string }
   */
  app.post('/api/onsong/send', async (request, reply) => {
    const token = process.env.ONSONG_TOKEN ?? '';
    if (token.length === 0) {
      return reply.status(503).send({ error: 'OnSong token not configured' });
    }

    const parsed = SendBodySchema.safeParse(request.body);
    if (!parsed.success) {
      return reply.status(400).send({ error: 'title, artist, and content are required' });
    }

    const { title, artist, content } = parsed.data;
    const filename = `${title} - ${artist}.txt`;

    // Build multipart/form-data using Node.js built-in FormData
    const form = new FormData();
    const blob = new Blob([content], { type: 'text/plain' });
    form.append('file[0]', blob, filename);

    let upstream: Response;
    try {
      upstream = await fetch(ONSONG_UPLOAD_URL, {
        method: 'POST',
        headers: {
          Authorization: token,
        },
        body: form,
      });
    } catch (err) {
      app.log.error(err, 'OnSong Cloud upload failed — network error');
      return reply.status(502).send({
        error: 'OnSong Cloud unreachable',
        detail: err instanceof Error ? err.message : 'Unknown error',
      });
    }

    if (!upstream.ok) {
      const body = await upstream.text().catch(() => '');
      app.log.error({ status: upstream.status, body }, 'OnSong Cloud upload returned non-2xx');
      return reply.status(502).send({
        error: `OnSong Cloud returned ${upstream.status}`,
        detail: body.slice(0, 500),
      });
    }

    return reply.send({ success: true, filename });
  });
}
